### PR TITLE
Support variable expansion in framework files

### PR DESCRIPTION
**Pull request by xuzhen**
1. Use ```safe_source``` to avoid [abuse of the framework files](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1019425#52)
2. Defer loading of signature related variables. #256 
3. Add tests.

### CODE REVIEW

**Comment by evelikov** [comment]
Do you mind splitting this in 3 patches as per the 1 - 2 - 3 points above. Thanks o/

**Comment by xuzhen** [comment]
> Do you mind splitting this in 3 patches as per the 1 - 2 - 3 points above. Thanks o/
OK. The commit has been split into smaller ones. New test cases are also split into individual commits

File: dkms.in [hunk 1]
```diff
@@ -875,10 +875,10 @@ prepare_signing()
                     sign_file="/usr/src/linux-headers-$kernelver/scripts/sign-file"
                 fi
                 ;;
-            * )
-                sign_file="/lib/modules/$kernelver/build/scripts/sign-file"
-                ;;
         esac
+        if [[ ! -f "${sign_file}" ]]; then
+            sign_file="/lib/modules/$kernelver/build/scripts/sign-file"
```
Comment (evelikov) [inline dkms.in:891]: Just a general note: my gut feeling is that we should remove all distribution logic out of dkms, in the long run.

File: dkms.in [hunk 2]
```diff
@@ -644,6 +645,13 @@ read_conf()
     return $return_value
 }
 
+# Source specified variables from dkms framework configuration files.
+read_framework_conf() {
+    for i in /etc/dkms/framework.conf /etc/dkms/framework.conf.d/*.conf; do
+        [[ -e "${i}" ]] && safe_source "$i" "$@"
```
Comment (evelikov) [inline dkms.in:651]: Nit: use `$i` in the check

File: dkms.in [hunk 3]
```diff
@@ -864,6 +872,9 @@ prepare_kernel()
 
 prepare_signing()
 {
+   # Lazy source in signing related configuration
+   read_framework_conf $dkms_framework_signing_variables
```
Comment (evelikov) [inline dkms.in:876]: Nit: quotes around the argument - here and below

Comment (xuzhen) [reply dkms.in:876]: `dkms_framework_signing_variables` is a string containing multiple variable names. `dkms_framework_signing_variables="sign_file mok_signing_key mok_certificate"` 
And `read_framework_conf` accepts multiple variable names as arguments. So there should be no quotation marks here.

Comment (evelikov) [reply dkms.in:876]: Right we're using `"$@"` in the function - I've misread that as `$1`

File: dkms.in (hunk 2 above)
Comment (xuzhen) [reply dkms.in]: Resolved in https://github.com/dell/dkms/pull/265/commits/5af2c9f4fe2686112c119ea2567b6a8b14b8738f

**Comment by evelikov** [comment]
@scaronni we might want to have a special note about [Fix framework file hijacking](https://github.com/dell/dkms/pull/265/commits/cf4040f7fe03d12717bd193e6507707384aa02ea) in the release notes.

it might break some people's workflow, but my gut feeling is that we should not revert it but gauge what they ~~hacks~~ use-case is.

**Pull request merged**